Source/CronosPlayUnreal/Private/DynamicContractActor.cpp

```diff
@@ -0,0 +1,241 @@
+    _coreContract = NULL;
+}
+
+void ADynamicContractActor::Destroyed() {
+    Super::Destroyed();
+
+    if (GEngine)
+        GEngine->AddOnScreenDebugMessage(
+            -1, 15.0f, FColor::Yellow,
+            TEXT("ADynamicContractActor  Destroyed"));
+
+    DestroyContract();
+
+    assert(NULL == _coreContract);
+}
+
+void ADynamicContractActor::DestroyContract() {
+    if (_coreContract != NULL) {
+        // restored back
+        rust::cxxbridge1::Box<org::defi_wallet_core::EthContract> tmpwallet =
+            rust::cxxbridge1::Box<org::defi_wallet_core::EthContract>::from_raw(
+                _coreContract);
+        _coreContract = NULL;
+    }
+}
+// Called when the game starts or when spawned
+void ADynamicContractActor::BeginPlay() { Super::BeginPlay(); }
+
+// Called every frame
+void ADynamicContractActor::Tick(float DeltaTime) { Super::Tick(DeltaTime); }
+
+void ADynamicContractActor::NewEthContract(FString cronosrpc,
+                                           FString contractaddress,
+                                           FString abijson, bool &success,
+                                           FString &output_message) {
+
+    try {
+        if (NULL != _coreContract) {
+            success = false;
+            output_message = TEXT("Contract Already Exists");
+            return;
+        }
+
+        std::string mycronosrpc = TCHAR_TO_UTF8(*cronosrpc);
+        std::string mycontract = TCHAR_TO_UTF8(*contractaddress);
+        std::string myjson = TCHAR_TO_UTF8(*abijson);
+
+        Box<EthContract> tmpContract =
+            new_eth_contract(mycronosrpc, mycontract, myjson);
+        // ownership transferred
+        _coreContract = tmpContract.into_raw();
+
+        success = true;
+    } catch (const rust::cxxbridge1::Error &e) {
+        success = false;
+        output_message =
+            FString::Printf(TEXT("CronosPlayUnreal NewEthContract Error: %s"),
+                            UTF8_TO_TCHAR(e.what()));
+    }
+}
+
+void ADynamicContractActor::NewSigningEthContract(
+    FString cronosrpc, FString contractaddress, FString abijson,
+    int32 walletindex, int64 chainid, bool &success, FString &output_message) {
+
+    try {
+        if (NULL != _coreContract) {
+            success = false;
+            output_message = TEXT("Contract Already Exists");
+            return;
+        }
+
+        if (NULL == defiWallet) {
+            success = false;
+            output_message = TEXT("Wallet Not Linked");
+            return;
+        }
+
+        assert(defiWallet != NULL);
+        if (NULL == defiWallet->_coreWallet) {
+            success = false;
+            output_message = TEXT("Wallet Not Initialized");
+            return;
+        }
+
+        char hdpath[100];
+        snprintf(hdpath, sizeof(hdpath), "m/44'/%d'/0'/0/%d", EthCoinType,
+                 walletindex);
+        rust::cxxbridge1::Box<PrivateKey> privatekey =
+            defiWallet->_coreWallet->get_key(hdpath);
```

Comment:
How to `get_key` if the we don't restore or create the wallet?

Reply:
it will return error, cannot proceed any more